### PR TITLE
Actions Artifacts support uploading multiple files and directories

### PR DESCRIPTION
**Pull request by fuxiaohei**
current actions artifacts implementation only support single file artifact. To support multiple files uploading, it needs:

- save each file to each db record with same run-id, same artifact-name and proper artifact-path
- need change artifact uploading url without artifact-id, multiple files creates multiple artifact-ids
- support `path` in download-artifact action. artifact should download to `{path}/{artifact-path}`.
- in repo action view, it provides zip download link in artifacts list in summary page, no matter this artifact contains single or multiple files.

### CODE REVIEW

**Review by lunny — APPROVED** [review]

File: routers/api/actions/artifacts_chunks.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	if written != artifact.FileCompressedSize {
+		return fmt.Errorf("merged file size is not equal to chunk length")
+	}
+
+	// save storage path to artifact
+	log.Debug("[artifact] merge chunks to artifact: %d, %s", artifact.ID, storagePath)
+	artifact.StoragePath = storagePath
+	artifact.Status = actions.ArtifactStatusUploadConfirmed
+	if err := actions.UpdateArtifactByID(ctx, artifact.ID, artifact); err != nil {
+		return fmt.Errorf("update artifact error: %v", err)
```
Comment (wxiaoguang) [inline routers/api/actions/artifacts_chunks.go:183]: Here and above (` written != artifact.FileCompressedSize `), does it need to delete the upload file if any error occurs ?

**Review by wxiaoguang — APPROVED** [review]

File: routers/web/repo/actions/view.go [hunk 2]
```diff
@@ -531,20 +525,49 @@ func ArtifactsDownloadView(ctx *context_module.Context) {
...
+	if len(artifacts) == 0 {
+		ctx.Error(http.StatusNotFound, "artifact not found")
+		return
+	}
 
-	ctx.ServeContent(f, &context_module.ServeHeaderOptions{
-		Filename:     artifact.ArtifactName,
-		LastModified: artifact.CreatedUnix.AsLocalTime(),
-	})
+	ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%s.zip; filename*=UTF-8''%s.zip", artifactName, artifactName))
```
Comment (silverwind) [inline routers/web/repo/actions/view.go]: `filename*` needs to be encoded via `url.PathEscape`, see `modules/context/context.go`. Also, there is some slash handling for `filename` there. I think we should make a shared function, used by both call sites.

Comment (silverwind) [reply routers/web/repo/actions/view.go]: Of course, this is assuming the filename can contain slashes or unicode characters, but still it's good to have the shared function nevertheless.

File: models/actions/artifact.go [hunk 3]
```diff
@@ -31,7 +31,7 @@ func init() {
 // ActionArtifact is a file that is stored in the artifact storage.
 type ActionArtifact struct {
 	ID                 int64 `xorm:"pk autoincr"`
-	RunID              int64 `xorm:"index UNIQUE(runid_name)"` // The run id of the artifact
```
Comment (wolfogre) [inline models/actions/artifact.go:34]: No migration for this?

**Review by silverwind — APPROVED** [review]

**Pull request merged**